[WIP] Add reference for network traffic (#30)
metalbear-co/mirrord.dev: content/docs/reference/traffic/index.md

---
title: "Network Traffic"
description: "Reference to working with network traffic with mirrord"
date: 2022-08-08T08:48:45+00:00
lastmod: 2022-08-08T08:48:45+00:00
draft: false
images: []
menu:
  docs:
    parent: "reference"
weight: 110
toc: true
---

## Incoming

> mirrord allows users to debug incoming network traffic in the following ways -

Reviewer: Suggestion: you could already name the two ways in this sentence. Right now the second option, "stealing" only comes later and there are a bunch of different other headings in between, so maybe it's not immediately clear that the two options are "mirroring" and "stealing".

> #### mirroring

Reviewer: Shouldn't this be capitalised?

|-------------|-------------------|--------------|---------------------------|
| default     | kubernetes        | No node port |
| default     | user-service      |           80 | http://192.168.49.2:32000 |
| kube-system | kube-dns          | No node port |
|-------------|-------------------|--------------|---------------------------|

mehula@mehul-machine:~/mirrord-demo$ curl -X POST -H "Content-type: application/json" -d "{\"Name\" : \"Metal\", \"Last\" : \"Bear\"}" http://192.168.49.2:32000/user
{"Last":"Bear","Name":"Metal"}

mehula@mehul-machine:~/mirrord-demo$ curl http://192.168.49.2:31000/index.html
<html> <head>USERS</head><body><h1> MetalBear Users</h1><p>[{"Last":"Bear","Name":"Metal"}]</p></body></html>
```

To mirror traffic from remote services to the local development environment, run the services locally with mirrord

```bash
mehula@mehul-machine:~/mirrord$ kubectl get pods
NAME                                        READY   STATUS    RESTARTS      AGE
metalbear-bff-deployment-597cb4f957-485t5   1/1     Running   1 (15h ago)   16h
metalbear-deployment-85c754c75f-6k7mg       1/1     Running   1 (15h ago)   16h

> ```

Reviewer: Maybe put the `kubectl get pods` snippet before the sentence that ends with `run the services locally with mirrord`, so what follows that sentence is running mirrord.

> Result -

Reviewer: I don't think this line is necessary, but it's up to you.

##### Window 1

```bash

> mehula@mehul-machine:~/mirrord-demo$ ../mirrord/target/debug/mirrord exec -c --pod-name metalbear-deployment-85c754c75f-6k7mg python3 user-service/service.py 

Reviewer: With https://github.com/metalbear-co/mirrord/pull/343 this would be outdated. Could you please create the examples using the new interface?

```suggestion
mehula@mehul-machine:~/mirrord-demo$ ../mirrord/target/debug/mirrord exec --no-outgoing -c --pod-name metalbear-deployment-85c754c75f-6k7mg python3 user-service/service.py 
```
Otherwise, you would get 2 responses, one from the remote process, one from the local process, right?
Or maybe it would be even more interesting to keep the command that way, but to show two responses. Whatever you think.

Author: @aviramha ptal
I have added `--no-outgoing` to be more precise, and i dont thinkk there'd be two responses. we are not requesting data over http in our sample service

Reviewer: hmm, we get a response though to the `curl` request, but actually you're right that there wouldn't be two lines in the shell, because the duplicate traffic is probably recognised as duplicate on the TCP level with the sequence number and ack number.
But you could just test it with outgoing traffic and see what happens :)

 * Serving Flask app 'service' (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on all addresses (0.0.0.0)
   WARNING: This is a development server. Do not use it in a production deployment.
 * Running on http://127.0.0.1:33695
 * Running on http://172.16.0.4:33695 (Press CTRL+C to quit)
 127.0.0.1 - - [08/Sep/2022 15:34:34] "GET /users HTTP/1.1" 200 - // <- Received mirrored traffic from the remote pod
```

##### Window 2

```bash
mehula@mehul-machine:~/mirrord-demo$ curl http://192.168.49.2:32000/users
[{"Last":"Bear","Name":"Metal"}]

> ```

Reviewer: Idea: you could split `Window 1` to 2 separate snippets and show the lines in chronological order. Maybe it would make it a bit easier for users to understand what's going on in this example.

Author: yep, that was the idea. looking for a nice way to do so in markdown

> #### stealing

Reviewer: Should this be capitalised?

> #### stealing

Reviewer 2: capital

> mirrord can steal network traffic, i.e. intercept it and send it to the local process. This means that the local process's state is directly interacting
> with the incoming network traffic without affecting the remote process.

Reviewer: I'm not sure about the 'process's state interacting with traffic' part. Maybe it could be clearer without mentioning state?
```suggestion
mirrord can steal network traffic, i.e. intercept it and send it to the local process. This means that all incoming 
traffic is only handled by the local process, not by the remote process.
```

##### Window 1

```bash
mehula@mehul-machine:~/mirrord-demo$ ../mirrord/target/debug/mirrord exec -c --tcp-steal --pod-name metalbear-deployment-85c754c75f-6k7mg python3 user-service/service.py 
 * Serving Flask app 'service' (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on all addresses (0.0.0.0)
   WARNING: This is a development server. Do not use it in a production deployment.
 * Running on http://127.0.0.1:35215
 * Running on http://172.16.0.4:35215 (Press CTRL+C to quit) 
 127.0.0.1 - - [08/Sep/2022 15:48:40] "GET /users HTTP/1.1" 200 -
 127.0.0.1 - - [08/Sep/2022 15:50:40] "POST /user HTTP/1.1" 200 -
 127.0.0.1 - - [08/Sep/2022 15:50:55] "GET /users HTTP/1.1" 200 -
 127.0.0.1 - - [08/Sep/2022 16:57:51] "POST /user HTTP/1.1" 200 -
 127.0.0.1 - - [08/Sep/2022 16:57:54] "GET /users HTTP/1.1" 200 -
 ^Cmehula@mehul-machine:~/mirrord-demo$ 
```

##### Window 2

```bash
// Before running mirrord with `--tcp-steal`
mehula@mehul-machine:~/mirrord-demo$ curl http://192.168.49.2:32000/users
[{"Last":"Bear","Name":"Metal"}]

> // After running with mirrord and `--tcp-steal` - remote process witched context with local process

Reviewer: typo: `witched`
Maybe saying that the processes switch contexts could be a bit confusing. They each keep their own context, it's just that now it's the local process that's answering and not the remote.

```suggestion
// When running with mirrord and `--tcp-steal`, we get a reply from the local process instead of the remote. 
```

mehula@mehul-machine:~/mirrord-demo$ curl http://192.168.49.2:32000/users
[]
mehula@mehul-machine:~/mirrord-demo$ curl -X POST -H "Content-type: application/json" -d "{\"Name\" : \"Mehul\", \"Last\" : \"Arora\"}" http://192.168.49.2:32000/user
{"Last":"Arora","Name":"Mehul"}
mehula@mehul-machine:~/mirrord-demo$ curl http://192.168.49.2:32000/users
[{"Last":"Arora","Name":"Mehul"}]
mehula@mehul-machine:~/mirrord-demo$ curl -X POST -H "Content-type: application/json" -d "{\"Name\" : \"Alex\", \"Last\" : \"C\"}" http://192.168.49.2:32000/user
{"Last":"C","Name":"Alex"}
mehula@mehul-machine:~/mirrord-demo$ curl http://192.168.49.2:32000/users
[{"Last":"Arora","Name":"Mehul"},{"Last":"C","Name":"Alex"}]

> // After sending SIGINT to the local process - remote switched back to its original context

Reviewer: Also here - I wouldn't say the remote switched context - it's just that it's the remote process getting the traffic and answering requests again.

---
title: "Network Traffic"
description: "Reference to working with network traffic with mirrord"
date: 2022-08-08T08:48:45+00:00
lastmod: 2022-08-08T08:48:45+00:00
draft: false
images: []
menu:
  docs:
    parent: "reference"
weight: 110
toc: true
---

> ## Incoming (TCP/UDP)

Reviewer 3: Since all features support TCP/UDP, no need to put it in the titles. Just add a footnote or something.

Author: added to the title

> mirrord allows users to debug incoming network traffic by mirroring and stealing.

Reviewer 3: ```suggestion
mirrord lets users debug incoming network traffic by mirroring or stealing the traffic sent to the remote pod.
```

#### Mirroring

> mirrord's default configuration is to duplicate traffic from the remote pod, i.e. run the local process in the context of cloud environment without

Reviewer 3: ```suggestion
mirrord's default configuration is to the mirror traffic that reaches the remote pod, i.e. run the local process in the context of cloud environment without
```

> manipulating incoming traffic.

Reviewer 3: ```suggestion
disrupting incoming traffic.
```

```

To mirror traffic from remote services to the local development environment, run the services locally with mirrord

<table>
<tr>
</tr>
<tr>
<td>
<center>Window 1</center> 

```bash
mehula@mehul-machine:~/mirrord-demo$ ../mirrord/target/debug/mirrord exec -c
--no-outgoing --pod-name metalbear-deployment-85c754c75f-6k7mg python3
user-service/service.py 
 * Serving Flask app 'service' (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on all addresses (0.0.0.0)
   WARNING: This is a development server. Do not use it in a production deployment.
 * Running on http://127.0.0.1:33695
 * Running on http://172.16.0.4:33695 (Press CTRL+C to quit)
 127.0.0.1 - - [08/Sep/2022 15:34:34] "GET /users HTTP/1.1" 200
// ^ Received mirrored traffic from the remote pod
```

</td>
<tr>
<td>
<center>Window 2</center>

```bash
mehula@mehul-machine:~/mirrord-demo$ curl http://192.168.49.2:32000/users
[{"Last":"Bear","Name":"Metal"}]
```

</td>
</tr>

</tr>
</table>

#### Stealing

> mirrord can steal network traffic, i.e. intercept it and send it to the local process. This means that all incoming

Reviewer 3: ```suggestion
mirrord can steal network traffic, i.e. intercept it and send it to the local process instead of the remote pod. This means that all incoming
```

```bash
// Before running mirrord with `--tcp-steal`
mehula@mehul-machine:~/mirrord-demo$ curl http://192.168.49.2:32000/users
[{"Last":"Bear","Name":"Metal"}]

// After running with mirrord and `--tcp-steal` - local process responds
 instead of the remote
mehula@mehul-machine:~/mirrord-demo$ curl http://192.168.49.2:32000/users
[]
mehula@mehul-machine:~/mirrord-demo$ curl -X POST -H 
"Content-type: application/json" 
-d "{\"Name\" : \"Mehul\", \"Last\" : \"Arora\"}" http://192.168.49.2:32000/user

{"Last":"Arora","Name":"Mehul"}
mehula@mehul-machine:~/mirrord-demo$ curl http://192.168.49.2:32000/users
[{"Last":"Arora","Name":"Mehul"}]
mehula@mehul-machine:~/mirrord-demo$ curl -X POST -H 
"Content-type: application/json" 
-d "{\"Name\" : \"Alex\", \"Last\" : \"C\"}" http://192.168.49.2:32000/user
{"Last":"C","Name":"Alex"}
mehula@mehul-machine:~/mirrord-demo$ curl http://192.168.49.2:32000/users
[{"Last":"Arora","Name":"Mehul"},{"Last":"C","Name":"Alex"}]

// After sending SIGINT to the local process
mehula@mehul-machine:~/mirrord-demo$ curl http://192.168.49.2:32000/users
[{"Last":"Bear","Name":"Metal"}]
```

</td>
</tr>

</tr>
</table>

## Outgoing (TCP/UDP)

> mirrord provides access to outgoing traffic, i.e. local network requests will be redirected to the remote.

Reviewer 3: ```suggestion
mirrord's outgoing traffic feature intercepts outgoing requests from the local process and sends them through the remote pod instead. Responses are then routed back to the local process.
```
